File: homework_iv/homework_iv_database/src/main/java/homework_iv/database/repository/MySqlAssistantRepository.java
package homework_iv.database.repository;

import homework_iv.database.model.Assistant;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

@Slf4j
public class MySqlAssistantRepository implements AssistantRepository {

    private static final String databaseUrl = "jdbc:mysql://localhost:3306/homework";
    private static final String username = "root";
    private static final String password = "my-secret-pw";

    private static final String nullableAssistant = "ALEKSANDAR";

    private static final String insertQuery = "insert into assistant_grades (name, score) values (?, ?)";

    @SneakyThrows
    public List<Assistant> getAssistantAverage() {
        log.info("Fetching data for assistant grading.");
        Class.forName("com.mysql.cj.jdbc.Driver");
        Connection connection = DriverManager.getConnection(databaseUrl, username, password);

        Statement statement = connection.createStatement();
        ResultSet resultSet = statement.executeQuery(
                "select name, sum(score) as score_sum, count(score) as grade_count from assistant_grades group by name");

        List<Assistant> assistantsGrading = new ArrayList<>();

        while(resultSet.next()) {
            int scoreSum = resultSet.getInt("score_sum");
            int gradeCount = resultSet.getInt("grade_count");

            assistantsGrading.add(
                    Assistant.builder()
                    .name(resultSet.getString("name"))
                    .averageScore(resultSet.getString("name").equals(nullableAssistant) ?
                            0 : ((float) scoreSum) / gradeCount)
                    .build()
            );
        }

        connection.close();

        log.info("Assistant summary: {}", assistantsGrading);

        return assistantsGrading;
    }

    @Override
    @SneakyThrows
    public void saveNewAssistant(Assistant assistant) {
        Class.forName("com.mysql.cj.jdbc.Driver");
        Connection connection = DriverManager.getConnection(databaseUrl, username, password);

        PreparedStatement preparedStatement = connection.prepareStatement(insertQuery);
        preparedStatement.setString(1, assistant.getName());
        preparedStatement.setInt(2, assistant.getScore());

        preparedStatement.execute();

        connection.close();
    }
}

File: homework_vi/src/main/java/rs/webprogramming/homework_vi/repository/MySqlAssistantRepository.kt
package rs.webprogramming.homework_vi.repository

import rs.app.povezi.util.Logger
import rs.webprogramming.homework_vi.model.Assistant
import java.sql.Connection
import java.sql.DriverManager

class MySqlAssistantRepository: AssistantRepository {

    companion object {
        private val host = "jdbc:mysql://localhost:3306/homework"
        private val username = "root"
        private val password = "my-secret-pw"

        private val nullableAssistant = "ALEKSANDAR"

        private val GET_AVERAGE_QUERY = "SELECT `name`, SUM(score) AS score_sum, COUNT(score) AS grade_count FROM assistant_grades GROUP BY name"
        private val INSERT_QUERY =
                "INSERT INTO assistant_grades (name, score) VALUES (?, ?)"
    }

    private fun openConnection(): Connection {
        Class.forName("com.mysql.cj.jdbc.Driver")
        var connection = DriverManager.getConnection(
                host, username, password
        )
        return connection
    }

    override fun getAssistantAverage(): MutableList<Assistant> {
        Logger.logInfo("Fetch data for assistant grading")
        val connection = openConnection()

        val statement = connection.createStatement()
        val resultSet = statement.executeQuery(GET_AVERAGE_QUERY)

        val assistantsGrading = mutableListOf<Assistant>()

        while(resultSet.next()) {
            val scoreSum = resultSet.getInt("score_sum")
            val gradeCount = resultSet.getInt("grade_count")

            var averageScore: Float

            averageScore = if(resultSet.getString("name") == nullableAssistant) {
                0F
            } else {
                (scoreSum.toFloat()) / gradeCount
            }

            assistantsGrading.add(
                    Assistant(
                            resultSet.getString("name"),
                            null,
                            averageScore
                    )
            )
        }

        connection.close()

        Logger.logInfo("Assistant summary: $assistantsGrading")

        return assistantsGrading
    }

    override fun saveNewAssistant(assistant: Assistant) {
        val connection = openConnection()

        val statement = connection.prepareStatement(INSERT_QUERY)
        statement.setString(1, assistant.name)
        statement.setInt(2, assistant.score!!)

        statement.execute()

        connection.close()
    }


}